### PR TITLE
PD-534: Support Segment Tracking calls in new navs

### PR DESCRIPTION
**Pull request by brookescarlett**
## ✍️ Proposed changes

- Definitely breaking some rules of `react-testing-library` which I am gathering based on the number of errors in the console when I run the test suite, but they all passed and everythign else works as expected so I figured I'd let y'all review while I clean up the tests

- Added testids throughout
- Added several elements to the NavElement enum and prefixed names for clarity
- Now every nav element that @nellem23 wanted to test should be available through the onElementClick function

🎟 _Jira ticket:_ [PD-534](https://jira.mongodb.org/browse/PD-534)

## 🛠 Types of changes

- [ ] Tooling (updates to workspace config or internal tooling)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## ✅ Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have run `yarn changeset` and documented my changes

### CODE REVIEW

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx [hunk 1]
```diff
@@ -224,7 +235,10 @@ function ProjectSelect({
       placeholder={current?.projectName ?? ''}
       open={open}
       loading={loading}
-      onClick={() => setOpen(curr => !curr)}
+      onClick={e => {
```
Comment (DesignerDave) [inline packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: This onClick doesn't have a typed event parameter like the others.

As an aside, you might consider making a wrapper for these repeated onClick handlers. Something like this:

```ts
type ClickHandler = (e: React.MouseEvent) => void

function getOnClickHandler(elementId: NavElement, callback?: ClickHandler): ClickHandler {
  return function(e) {
    onElementClick(elementId)

    if (callback) {
      callback(e)
    }
  }
}
```

Then you can use it in most cases like:
```js
onClick={getOnClickHandler(NavElement.ProjectNavViewAllProjects)}
```
And in cases like this, you can do:
```js
onClick={getOnClickHandler(NavElement.ProjectNavProjectSelectTrigger, () => setOpen(curr => !curr))}
```

File: packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+    expectedElements[NavElement.OrgNavDropdown] = queryByTestId(
+      'org-nav-dropdown',
+    );
+    expectedElements[NavElement.OrgNavDropdownOrgAccessManager] = queryByTestId(
+      'org-nav-dropdown-org-access-manager',
+    );
+    expectedElements[
+      NavElement.OrgNavDropdownProjectAccessManager
+    ] = queryByTestId('org-nav-dropdown-project-access-manager');
+    expectedElements[NavElement.OrgNavSupport] = queryByTestId(
```
Comment (DesignerDave) [inline packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx]: This is a lot of repeat logic. Couldn't you do something like this?
```js
const expectedElements = Object.keys(NavElement).reduce((acc, val) => {
  acc[val] = queryByTestId(NavElement[val])
  return acc
}, {} as ExpectedElements)
```

File: packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,273 @@
+import React from 'react';
+import { render, cleanup, act, fireEvent } from '@testing-library/react';
+import '@testing-library/jest-dom/extend-expect';
+import { nullableElement, Queries } from '@leafygreen-ui/lib/src/testHelpers';
+import MongoNav from '../MongoNav';
+import { NavElement } from '../types';
```
Comment (DesignerDave) [inline packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx]: Could we add a test to ensure all values are unique in the NavElement object?

File: packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx (hunk 2 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx]: Had to rewrite test suite, still not super clean but lmk what you think now... significantly decreased the warnings in the console..

File: packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx (hunk 3 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx]: 👍

File: packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx (hunk 2 above)
Comment (DesignerDave) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClick.spec.tsx]: Seems significantly cleaner than before 👍

File: packages/mongo-nav/src/mongo-select/MongoSelect.tsx (hunk 1 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/mongo-select/MongoSelect.tsx]: Spoke off-line, leaving as-is

**Comment by brookescarlett** [comment]
@hswolff @DesignerDave do we want to release this as MongoNav v1 once it's in?

File: .changeset/hip-gorillas-joke.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+---
```
Comment (hswolff) [inline .changeset/hip-gorillas-joke.md]: Need additional changeset for segment tracking calls.

File: packages/mongo-nav/src/MongoNav.story.tsx [hunk 5]
```diff
@@ -24,17 +24,18 @@ storiesOf('MongoNav', module).add('Default', () => (
         activeNav={select(
           'activeNav',
           Object.values(NavElement),
```
Comment (hswolff) [inline packages/mongo-nav/src/MongoNav.story.tsx]: Hrm....this is now a lie. `NavElement` contains way more than what you can actually set as an activeNav value.

We should probably make a separate object that just contains the values that we'll actually use here, like `ActiveNav` as the object and its values can be taken from `NavElement`.

File: packages/mongo-nav/src/org-nav/OrgNav.tsx [hunk 6]
```diff
@@ -243,75 +246,95 @@ export default function OrgNav({
...
+            Access Manager
+          </OrgNavLink>
+
+          <IconButton
+            ariaLabel="Dropdown"
+            active={accessManagerOpen}
+            disabled={!current}
+            data-testid="org-nav-dropdown"
+            onClick={(e: React.MouseEvent) => {
+              onElementClick(NavElement.OrgNavDropdown, e);
```
Comment (hswolff) [inline packages/mongo-nav/src/org-nav/OrgNav.tsx]: So here's an idea of a refactor for the value we get back from `useOnElementClick`.

What if we wrap the value given as a prop into a function that looks like this:

```jsx
const wrappedOnElementClick = (navElement: NavElement, additionalWork: (e) => void) {
   return (e) => {
        onElementClick(NavElement.OrgNavDropdown, e);
        return additionalWork(e);
  };
};

// So you could use it one of these two ways:
<MyComponent onClick={onElementClick(NavElement.OrgNavDropdown)} />
// or
<MyComponent onClick={onElementClick(NavElement.OrgNavDropdown, () => setStateThing())} />
```

File: .changeset/hip-gorillas-joke.md (hunk 4 above)
Comment (brookescarlett) [reply .changeset/hip-gorillas-joke.md]: I know but wasn't sure if we wanted this to be major as our first official release, or if we were going to continue with minor updates. What do you think?

File: packages/mongo-nav/src/org-nav/OrgNav.tsx (hunk 6 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/org-nav/OrgNav.tsx]: 👍 woo!

File: .changeset/hip-gorillas-joke.md (hunk 4 above)
Comment (hswolff) [reply .changeset/hip-gorillas-joke.md]: I think we should hold off. I'm going to have a few more things to change from integrating this into cloud.

I think once we go live on cloud-dev we can bump this to 1.0.0.

Comment (brookescarlett) [reply .changeset/hip-gorillas-joke.md]: 👍

File: packages/mongo-nav/src/MongoNav.story.tsx (hunk 5 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/MongoNav.story.tsx]: 👍 Didn't construct the most beautifully but it does work

File: packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+export function useOnElementClick() {
+  return useContext(OnElementClickContext);
+}
+
+export default function OnElementClickProvider({
+  children,
+  onElementClick,
+}: OnElementClickProviderProps) {
+  const wrappedOnElementClick = (
```
Comment (hswolff) [inline packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]: Need this memoized. Your usage of `useMemo` below requires the second argument: `useMemo(functionToMemo, dependencyArrayLikeUseEffect)`.

So for example `useMemo(wrappedOnElementClick, [onElementClick])`

File: packages/mongo-nav/src/types.ts [hunk 8]
```diff
@@ -1,18 +1,50 @@
+const ActiveNavElement = {
+  OrgNavOrgSettings: 'orgNavOrgSettings',
```
Comment (hswolff) [inline packages/mongo-nav/src/types.ts]: We should at least do 
```suggestion
  OrgNavOrgSettings: NavElement.OrgNavOrgSettings,
```
etc

File: packages/mongo-nav/src/types.ts [hunk 9]
```diff
@@ -1,18 +1,50 @@
+const ActiveNavElement = {
```
Comment (hswolff) [inline packages/mongo-nav/src/types.ts]: We should update all usages of `activeNav === XX` to use this object instead, as that's the object that dictates what in the nav can be set as active.

File: packages/mongo-nav/src/MongoNav.tsx [hunk 10]
```diff
@@ -275,6 +218,9 @@ function MongoNav({
       orgSettings: `${hosts.cloud}/v2#/org/${data?.currentOrganization?.orgId}/settings/general`,
     },
     orgNav: {
+      leaf: data?.currentOrganization
+        ? `${hosts.cloud}/v2#/org/${data?.currentOrganization?.orgId}/`
```
Comment (hswolff) [inline packages/mongo-nav/src/MongoNav.tsx]:
```suggestion
        ? `${hosts.cloud}/v2#/org/${data.currentOrganization?.orgId}/`
```

Can drop that one as you checked it in the ternary. It'd be nice to not also have to check if `currentOrganization` exists, and do that in the first part of the ternary too.

File: packages/mongo-nav/src/user-menu/UserMenuTrigger.tsx [hunk 11]
```diff
@@ -99,10 +101,13 @@ export default function UserMenuTrigger({
   children,
   ...rest
 }: UserMenuTriggerProps) {
+  const onElementClick = useOnElementClick();
   const buttonRef = useRef<HTMLButtonElement>(null);
   const { width: viewportWidth } = useViewportSize();
   const isTablet = viewportWidth < breakpoints.medium;
 
+  // console.log('open is', open, onElementClick);
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/mongo-nav/src/user-menu/UserMenuTrigger.tsx]: 👍

File: packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx (hunk 7 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]: 👍

File: packages/mongo-nav/src/types.ts (hunk 8 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/types.ts]: 👍

File: packages/mongo-nav/src/types.ts (hunk 9 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/types.ts]: 👍

File: packages/mongo-nav/src/MongoNav.tsx (hunk 10 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/MongoNav.tsx]: Isn't this checking that currentOrg exists in the first part of the ternary?

File: packages/mongo-nav/src/MongoNav.tsx [hunk 12]
```diff
@@ -219,7 +220,7 @@ function MongoNav({
     },
     orgNav: {
       leaf: data?.currentOrganization
-        ? `${hosts.cloud}/v2#/org/${data?.currentOrganization?.orgId}/`
+        ? `${hosts.cloud}/v2#/org/${data.currentOrganization?.orgId}/`
```
Comment (hswolff) [inline packages/mongo-nav/src/MongoNav.tsx]:
```suggestion
        ? `${hosts.cloud}/v2#/org/${data.currentOrganization.orgId}/`
```

Should also be valid then?

File: packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx [hunk 13]
```diff
@@ -37,29 +37,12 @@ export default function OnElementClickProvider({
...
-  //     return (e: React.MouseEvent) => {
-  //       onElementClick(navElement, e);
-  //       if (additionalWork) {
-  //         return additionalWork(e);
-  //       }
-  //     };
-  //   },
-  //   [onElementClick],
-  // );
+  const memoizedWrapper = React.useMemo(() => wrappedOnElementClick, [
```
Comment (hswolff) [inline packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]:
```suggestion
  const memoizedWrapper = React.useMemo(wrappedOnElementClick, [
```

`useMemo` is like `debounce`, it passes on whatever arguments are given to the function returned from memo.

File: packages/mongo-nav/src/MongoNav.tsx (hunk 12 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/MongoNav.tsx]: 👍

File: packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx (hunk 13 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]: 👍

File: packages/mongo-nav/src/project-nav/ProjectNav.tsx [hunk 14]
```diff
@@ -289,55 +295,53 @@ export default function ProjectNav({
...
           >
             <Icon glyph="Ellipsis" className={menuIconStyle} />
           </IconButton>
         }
       >
         <MenuItem
           href={projectNav.settings}
-          onClick={onMenuClick}
           data-testid="project-nav-settings"
+          onClick={onElementClick(ProjectNavProjectSettings, () =>
```
Comment (hswolff) [inline packages/mongo-nav/src/project-nav/ProjectNav.tsx]: FYI could have also done: 
```suggestion
          onClick={onElementClick(ProjectNavProjectSettings, onMenuClick)}
```

File: packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx [hunk 15]
```diff
@@ -0,0 +1,51 @@
...
+      onElementClick(navElement, e);
+      return additionalWork?.(e);
+    };
+  };
+
+  const callbackWrapper = useCallback(
+    (navElement, additionalWork) =>
+      wrappedOnElementClick(navElement, additionalWork),
+    [onElementClick],
+  );
```
[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]: 👍

[the remark this answers is not in the document]
Comment (brookescarlett) [reply packages/mongo-nav/src/on-element-click-provider/OnElementClickProvider.tsx]: 👍

File: packages/mongo-nav/src/project-nav/ProjectNav.tsx (hunk 14 above)
Comment (brookescarlett) [reply packages/mongo-nav/src/project-nav/ProjectNav.tsx]: I can revert if you prefer, lmk

Comment (hswolff) [reply packages/mongo-nav/src/project-nav/ProjectNav.tsx]: Nah, up to you, just letting you know alternative ways to solve the problem.

**Pull request merged**